### ISSUE TITLE
Dynamic routes cause a "ChunkLoadError" when using the app router in a dockerized Next.js 13 SPA

### ISSUE THREAD

**Issue by eide-1** [vercel/next.js#54008]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: linux
      Arch: x64
      Version: #85-Ubuntu SMP Fri Jul 7 15:25:09 UTC 2023
    Binaries:
      Node: 18.15.0
      npm: 9.5.0
      Yarn: N/A
      pnpm: N/A
    Relevant Packages:
      next: 13.4.15
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
      typescript: N/A
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

App Router

### Link to the code that reproduces this issue or a replay of the bug

https://github.com/eide-1/sample-project

### To Reproduce

**Steps to Reproduce:**
1. Run the the project in a production environment by following the instructions in the README file
2. Click on the "Go to Dynamic Route" button. You will be presented with an application error
3. Right-click anywhere on the webpage and click "Inspect". The browser console will display the "ChunkLoadError"

### Describe the Bug

I am currently working on upgrading an existing Next.js SPA that uses the pages router to the app router. The project is deployed using Docker and Nginx. However, I noticed that dynamic routes cause a "ChunkLoadError" when the SPA is deployed in a production environment. All pages (including dynamic routes) are functional in a development environment. 

The dynamic route in question uses the "use client" directive. I noticed that if I remove the "use client" directive and any component or hook that requires it from the page, I can get it to render properly in production mode.

I created a sample project that uses the same tech stack as my current SPA to demonstrate the issue and I included the steps to reproduce the issue below.  

I also found the following Stack Overflow question detailing a similar issue to the one I described above: [In Next.js 13 using App Router, why can't I export dynamic routes with "use client"?](https://stackoverflow.com/questions/76824224/in-next-js-13-using-app-router-why-cant-i-export-dynamic-routes-with-use-clie)

### Expected Behavior

Dynamic routes should work in the production environment just like they used to under the pages router without causing a ChunkLoadError.

### Which browser are you using? (if relevant)

Microsoft Edge 115.0.1901.203 (64-bit)

### How are you deploying your application? (if relevant)

Docker/Nginx

<sub>[NEXT-1641](https://linear.app/vercel/issue/NEXT-1641/dynamic-routes-cause-a-chunkloaderror-when-using-the-app-router-in-a)</sub>

**Comment by hungcrush** [comment]
I tested on my Local (next build & next start) it works well. But I don't know why on my server it not works.

**Comment by danielle-o3h** [comment]
I am seeing the same issue in our Docker Environments reverting to version 13.4.9 is a temporary fix for the issue.

**Comment by hykelvinlee42** [comment]
> I am seeing the same issue in our Docker Environments reverting to version 13.4.9 is a temporary fix for the issue.
> 
> Any urls with [] (%5B, %5D) return a 404 error which includes the generated js files for any dynamic page

I can confirm this solution works. The issue begins occurring since version `13.4.13-canary.16`.

**Comment by harryoppa** [comment]
Anyone can check this??? :(( I can not use newest version..

**Comment by Baukaalm** [comment]
I have the same issue, described it here https://github.com/vercel/next.js/issues/54133#issuecomment-1683366199.

**Comment by bryceAebi** [comment]
@hungcrush perhaps this is a clue as to why it's working locally but not on prod? Some issue with the caching of chunks. FWIW, this workaround did not solve the issue for my dynamic routes

https://github.com/vercel/next.js/issues/38507#issuecomment-1680749972

**Comment by bryceAebi** [comment]
I have more or less the same setup and the same issue, but I am only experiencing the error when I have Google Search Console "live test" my URL, which makes this difficult to debug.

I'm not able to reproduce the error in production via Chrome Beta.

**Comment by scottpaulin** [comment]
I am experiencing similar issues with Google Search Console "live test" on production URL for a site deployed on Vercel. It makes it difficult to get pages indexed in Google Search.

It seems a little non-deterministic for me. Sometimes Google Search Console "live test" does not work on a page. Yesterday  Google Search Console "live test" failed twice on a page then worked the third time.

next@13.4.4 and app router deployed on Vercel

**Comment by bryceAebi** [comment]
@scottpaulin I realized that my issue is not limited to my dynamic routes. It affects a bunch of my pages. The Live Test fails about two thirds of the time. But I do think my pages are getting indexed as I can find them using Google Search and my production pages load just fine. Perhaps it's an issue with the Live Test crawler erroneously thinking that a certain chunk is unnecessary. I'm hoping it's not actually impacting SEO...

**Comment by scottpaulin** [comment]
@bryceAebi yea I have seen it in a static route too.

Google also (eventually) indexes my pages. I think Google is recrawling pages every few days and eventually has a successful render.

Ahrefs is showing lots of errors about failed metadata etc, I'm guessing it's because the render fails.

Yea not sure if seo is being affected. Fingers crossed it isn't. Pretty scary issue for sites that rely on search traffic for revenue. 

I am seeing a bunch of http status 0 errors in datadog. Might be unrelated though.

**Comment by marwinlewis** [comment]
In my case I have dynamic routing pages > [category] > index.tsx and [id].tsx. The problem started after upgrading nextjs. I also tried the router app and the same problem. 404 in compiled js file in production environment. Downgrading nextjs to 13.3.4 fixed this issue for me.

**Comment by LucaNerlich** [comment]
ive tested many versions and `13.4.12` is the last working one. As soon as I upgrade next to any version higher, dynamic routing breaks in interesting ways.

404 ChunkLoadError in a Docker Environment and just no content and no error for non docker deployments.
Super weird and very annoying. Also only on unix? I am unable to reproduce this locally - with any version. Even latest works fine. But as soon as I deploy on a ubuntu cloud server -> router is broken again.

This affects both app and pages router.

**Comment by marwinlewis** [comment]
> ive tested many versions and `13.4.12` is the last working one. As soon as I upgrade next to any version higher, dynamic routing breaks in interesting ways.
> 
> 404 ChunkLoadError in a Docker Environment and just no content and no error for non docker deployments. Super weird and very annoying. Also only on unix? I am unable to reproduce this locally - with any version. Even latest works fine. But as soon as I deploy on a ubuntu cloud server -> router is broken again.
> 
> This affects both app and pages router.

Yes, locally even the production version works fine for the latest version. The problem occurs only after deployment to a unix server.

**Comment by shehi** [comment]
I am seeing this issue on PROD builds, starting v13.4.13 (works fine in v13.4.12). Testing every single release from v13.4.13 to v13.4.19 - all fail.

Dev env works just fine on the latest NextJs.

**Comment by domosedov** [comment]
same problem after update next.js 13.4.13+

**Comment by domosedov** [comment]
next.js 13.4.12 works fine
<img width="812" alt="Снимок экрана 2023-08-28 в 12 31 20" src="https://github.com/vercel/next.js/assets/42298502/730a6ade-05e6-4232-8cbf-8275d43b4edc">

next.js 13.4.13 + broke my app
<img width="665" alt="Снимок экрана 2023-08-28 в 12 30 42" src="https://github.com/vercel/next.js/assets/42298502/37cd0c0d-c946-40fd-8404-aa44aedc1f3d">

**Comment by pwestern** [comment]
For what it's worth, I have also had the same issue, and downgrading to `13.4.12` resolves it.

**Comment by shehi** [comment]
Coupla days ago while checking 13.4.20-canary release notes I noticed a fix
there. Haven't tested them myself of course (because of @sentry packages
requiring proper nextjs versions), but I think the fix is coming.

**Comment by janpaepke** [comment]
Guys this caused me a lot of head scratching. Thanks for this issue!

For the record (host OS: OSX):
- local dev env (`npm run dev`) **works** 
- local build start (`npm run build && npm start`) **works**
- local dockerized run (docker build, start container, forward port) **WORKS**!
- deployed dockerized run - **BREAKS**

The most confusing part to me was that even running in docker worked for me locally.
It lead me down a rabbit hole investigating if there are some weird ingress configurations on our cluster.

Whenever I deployed it, on the production version it would show:

"Application error: a client-side exception has occurred (see the browser console for more information)."
In the console I could then see that all the js chunks that are in dynamic routes could not be loaded.
I even checked in the docker container to confirm that the files are there (`.next/static/chunks/app/[dynamicBit]`).

It turned out to be this.
- Downgrading to 13.4.12 **fixed this**
- Upgrading to 13.4.20-canary.18 did **NOT fix this.**

Hope this saves someone some head-scratching.

I will investigate some more what might be going on here.

**Comment by domosedov** [comment]
13.4.12 - Latest stable version

**Comment by janpaepke** [comment]
I have further narrowed down, that the breaking change must have been introduced in [v13.4.13-canary.0](https://github.com/vercel/next.js/releases/tag/v13.4.13-canary.0) – it's the first version where this occurs.

Looking at the changes, nothing stood out and unfortunately I lack the time for further analysis.
Maybe this is a good starting point for someone, though.

Until then, we'll (have to) stick with 13.4.12

**Comment by poorscousertommy8** [comment]
same issue [here](https://github.com/vercel/next.js/issues/54325)
I also hope that the Next.js team is working on it. We are currently stagnating on version 13.4.12

**Comment by beverloo** [comment]
I've also been having this issue, and can reproduce Jan's observation that this broke between v13.4.12 and v13.4.13-canary.0, which contains the following commits:

https://github.com/vercel/next.js/compare/v13.4.12...v13.4.13-canary.0

Another observation is the following: given a dynamic route parameter **[slug]**,

- "static/chunks/app/events/[slug]/layout-dd5472667df98622.js" exists on disk and is included in the build manifest,
- "/_next/static/chunks/app/events/**%5Bslug%5D**/layout-dd5472667df98622.js" is requested by the browser, but 404s,
- "/_next/static/chunks/app/events/**%255Bslug%255D**/layout-dd5472667df98622.js" works just fine.

Since the file exists on disk without any form of encoding, which also was the case in v13.4.12, that makes me think this likely is caused by a change in the included server.

[This commit](https://github.com/vercel/next.js/commit/1398de9977b89c9c4717d26e213b52bc63ccbe7e) changes to how files are resolved, particularly the diffs in `packages/next/src/server/lib/router-utils/`. I see some encoding and decoding going on in there, but nothing stands out as the cause. There's an interesting case in `filesystem.ts` lines 560-565 that may explain why this isn't broken in dev?

```typescript
// In dev, we ensure encoded paths match
// decoded paths on the filesystem so check
// that variation as well
const tempItemPath = decodeURIComponent(curItemPath)
fsPath = path.posix.join(itemsRoot, tempItemPath)
found = await fileExists(fsPath, FileType.File)
```

I'm not really having any luck getting local changes to end up in the docker image, will try that again later...

**Comment by beverloo** [comment]
I've confirmed that https://github.com/vercel/next.js/commit/1398de9977b89c9c4717d26e213b52bc63ccbe7e is the cause of the issue.

Looking at `packages/next/src/server/lib/router-utils/filesystem.ts`, the check on line 517 fails because `curItemPath` is not found in `items` (for type=`nextStaticFolder`). The state at this point is:

```javascript
let curItemPath = "/_next/static/chunks/app/events/[slug]/page-8f8f16ccd387e59d.js";
let curDecodedItemPath = "/_next/static/chunks/app/events/[slug]/page-8f8f16ccd387e59d.js";
let items = [
    // ...
    "/_next/static/chunks/app/events/%5Bslug%5D/page-8f8f16ccd387e59d.js",
];
```

The entry in `items` was encoded on line 181:
```typescript
// ensure filename is encoded
nextStaticFolderItems.add(
  path.posix.join('/_next/static', encodeURI(file))
)
```

If I remove the `encodeURI` call the app works properly and this issue is solved. I see that files in the public directory are similarly encoded on :147; I can reproduce this exact issue in that directory using a file called `[hello].txt`, which 404s in v13.4.13-canary.0, but works when I remove that `encodeURI` call too.

Hey @ijjk - I hope you don't mind the mention - are you sure that the `encodeURI` calls are necessary here? Iow. could it be that the change you made to `test/integration/i18n-support/test/shared.js` hides an actual failure?

**Comment by beverloo** [comment]
Final message from me, apologies for triple posting. While this definitely is a regression in NextJS, I found the underlying root cause for my case: nginx was messing with the URL.

I use nginx to terminate TLS and reverse proxy the requests. When the `proxy_pass` rule ends with a slash, it gets rewritten. When it doesn't end with a slash, the original request URL is passed through. (Obviously 🙄...)

I fixed it by changing my location rule from this:
```
location / {
	proxy_pass http://localhost:3001/;  # <-- ends with a slash
	proxy_set_header Host $host;
	proxy_set_header X-Real-IP $remote_addr;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
}
```

to the following, after which the latest NextJS canary works just fine:
```
location / {
	proxy_pass http://localhost:3001;  # <-- does not end with a slash
	proxy_set_header Host $host;
	proxy_set_header X-Real-IP $remote_addr;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
}
```

**Comment by shehi** [comment]
> Final message from me, apologies for triple posting. While this definitely is a regression in NextJS, I found the underlying root cause for my case: nginx was messing with the URL.
> 
> I use nginx to terminate TLS and reverse proxy the requests. When the `proxy_pass` rule ends with a slash, it gets rewritten. When it doesn't end with a slash, the original request URL is passed through. (Obviously 🙄...)
> 
> I fixed it by changing my location rule from this:
> 
> ```
> location / {
> 	proxy_pass http://localhost:3001/;  # <-- ends with a slash
> 	proxy_set_header Host $host;
> 	proxy_set_header X-Real-IP $remote_addr;
> 	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
> }
> ```
> 
> to the following, after which the latest NextJS canary works just fine:
> 
> ```
> location / {
> 	proxy_pass http://localhost:3001;  # <-- does not end with a slash
> 	proxy_set_header Host $host;
> 	proxy_set_header X-Real-IP $remote_addr;
> 	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
> }
> ```

This indeed works quite fine. Upgraded to the latest Next.js and no dynamic route asset issues anymore. Thank you @beverloo !

**Comment by poorscousertommy8** [comment]
Unfortunately the problem still exists in v13.4.20-canary.23. In production mode, the browser immediately throws a 404 error and the entire website crashes as soon as I access it. When I open the website locally in production mode, no error occurs. Only on a Windows IIS server with iisnode.
It's starting to get worrying for me, we recently switched to NextJS (at my will) and it seems like the problem isn't going away :(

@beverloo do you think the problem needs to be fixed by the Next.js team or do we also need to work on a server solution?

The structure of my app (with my dynamic route) is as follows:
src -> app -> [...pagename] -> page.tsx

**Comment by janpaepke** [comment]
Maybe you could shed some light on this @ijjk, since you're the one who introduced the change that @beverloo has identified as the root cause here?
What was the idea behind the revert and what tests should we run in order to confirm wether or not @beverloo's [fix](https://github.com/beverloo/next.js/commit/abe85b7ccd14925d4a828850b6a0f989678e6f62) is viable?

**Comment by beverloo** [comment]
> Maybe you could shed some light on this @ijjk, since you're the one who introduced the change that @beverloo has identified as the root cause here? What was the idea behind the revert and what tests should we run in order to confirm wether or not @beverloo's [fix](https://github.com/beverloo/next.js/commit/abe85b7ccd14925d4a828850b6a0f989678e6f62) is viable?

For what it's worth, I don't think my proposed fix is correct anymore. My guess would be that this would be better as input validation earlier in the request handling - perhaps something like encoding the request URL if unencoded characters are found.

**Comment by rihards-simanovics** [comment]
> I fixed it by changing my location rule from this:
> 
> ```
> location / {
> 	proxy_pass http://localhost:3001/;  # <-- ends with a slash
> 	proxy_set_header Host $host;
> 	proxy_set_header X-Real-IP $remote_addr;
> 	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
> }
> ```
> 
> to the following, after which the latest NextJS canary works just fine:
> 
> ```
> location / {
> 	proxy_pass http://localhost:3001;  # <-- does not end with a slash
> 	proxy_set_header Host $host;
> 	proxy_set_header X-Real-IP $remote_addr;
> 	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
> }
> ```

Hey @beverloo, this seems to actually fix it for me, so weird. My stack is a self-hosted ubuntu 22.04, nginx, pm2, node 18.17.1, and now latest nextjs 13.4.19.

**Comment by janpaepke** [comment]
unfortunately this solution doesn't really help me, as we don't use nginx but use the standalone nextjs server.

**Comment by omarmciver** [comment]
Same issue here. Use of Dynamic routes produce 404 with next start, but work fine in next dev. Downgrade to 13.4.12 works, 13.4.13+ (up to .19) breaks.

**Comment by omarmciver** [comment]
Funny though, I have a deployment from 14 days ago (August 30th) which has next ^13.4.17 and that works fine. Going to play with other differences that might be affecting it.

My initial thought is webpack. We added webpack-merge since last time. Will update if I find anything.

**Comment by AmirL** [comment]
@omarmciver Thanks! Proxy pass didn't work correctly for 13.4.1, but works well for 13.4.12

location /_next {
  proxy_pass http://127.0.0.1:800;
  ...

**Comment by AmirL** [comment]
Hm, a strange thing. It works well for a single dynamic parameter, but doesn't work with 2 parameters. 
curl -I https://.../_next/static/chunks/app/%5Blang%5D/organization/%5BcustomerId%5D/edit/layout-88c4c469c2cd0946.js - status 301, do decoded url, which is not found. 

At the same time
curl -I https://.../_next/static/chunks/app/%5Blang%5D/organization/overview/layout-905473effe8e446d.js
returns 200, without any redirects. 

And edit page works well, if I go there from a working page, but doesn't work when I refresh browser on the page. 

Tested with versions 13.4.12 and 13.4.19

**Comment by AmirL** [comment]
I assumed that the issue from the previous message was related to different letter case in characters, and it turned out to be indeed the case. URLs work correctly when using [id] or [customer_id], but they do not work when using [customerId]. Otherwise, version 13.4.19 with proxy pass without a trailing slash works correctly.

**Comment by w3cproject** [comment]
tested with `v13.4.20-canary.35` - doesn't work.
`13.4.12` - works perfectly!

**Comment by janpaepke** [comment]
Current summary:
This regression concerns all versions since [v13.4.13-canary.0](https://github.com/vercel/next.js/releases/tag/v13.4.13-canary.0)
@beverloo [previously identified](https://github.com/vercel/next.js/issues/54008#issuecomment-1712309065) commit https://github.com/vercel/next.js/commit/1398de9977b89c9c4717d26e213b52bc63ccbe7e as the cause of the issue, which means it happened during this PR: https://github.com/vercel/next.js/pull/52492.
Users running nginx [can get around this issue](https://github.com/vercel/next.js/issues/54008#issuecomment-1712672092) by removing the trailing slash in their proxy_pass rule.
However this issue remains unresolved for users using next start or the standalone server.

I'd love to offer my help but I'd need some insights from @ijjk, since the PR and commit mentioned above are quite extensive, so it's hard to pinpoint the potential origin.

**Comment by poorscousertommy8** [comment]
https://github.com/vercel/next.js/issues/54008#issuecomment-1725005598

Maybe @timneutkens can help us with the problem. Are you aware of the problem? 
_(I waited a long time to mention you and checked the latest releases every day, but unfortunately the problem has persisted for weeks)_

There are a few other open issues that I think describe exactly the same problem.

**Comment by AmirL** [comment]
@janpaepke There is also a problem with different letter case, "[slug]" works, but [slugCase] doesn't.

**Comment by shehi** [comment]
Camelcase with Nginx fix works fine for me.

**Comment by LucaNerlich** [comment]
changing the nginx config is **not** a fix - this is at max a hacky workaround

**Comment by poorscousertommy8** [comment]
> @janpaepke There is also a problem with different letter case, "[slug]" works, but [slugCase] doesn't.

my slug is [...pagename]
Unfortunately does not work.

**Comment by shehi** [comment]
Nginx fix is actually a fix in real sense, as most of us misconfigure our
root URLs adding trailing slashes to them which cause URL rewrite. As I
mentioned earlier, some frameworks also counter that - e.g. Laravel (by
that extension, Symfony) is one such: put a trailing slash in your API
tests and watch redirects. By fixing Nginx config, we make sure that
doesn't happen in Nextjs.

As to Next's standalone server: apparently it has its own URL rewrite
issues, which have surfaced here. Nginx fix doesn't fix it of course, but
it fixes our use case, where Next server is replaced by Nginx. I hope I
could explain my point.

**Comment by janpaepke** [comment]
> As to Next's standalone server: apparently it has its own URL rewrite
issues, which have surfaced here. Nginx fix doesn't fix it of course, but
it fixes our use case, where Next server is replaced by Nginx. I hope I
could explain my point.

@shehi yes but I'd like to focus on the actual issue which is an obvious regression.

> There is also a problem with different letter case, "[slug]" works, but [slugCase] doesn't.

@AmirL , the same as @poorscousertommy8 I have this issue even with all lowercase, [lang] to be specific.
This is implementing i18n with the app router, roughly following [the official example](https://github.com/vercel/next.js/tree/canary/examples/app-dir-i18n-routing).

We have mentioned the issue to several maintainers now and it is clear we need their support.
Let's make it more likely that they will look into this issue by trying to limit additional comments to those which might shed further light on a potential solution.

**Comment by poorscousertommy8** [comment]
Doesn't work with update 13.5.1 either.
If I start it locally (production build), then it **works**, it tries to load
/_next/static/chunks/app/page-a599bfbc7c8a608a.js

If I use exactly the same build on the server (Windows), then it tries to load a completely different file (encoded):
/_next/static/chunks/app/%5B...pagename%5D/page-8b669ad10c44488e.js
which doesnt work.
Why is that? I do not get it.

**Comment by bielinskiadrian** [comment]
Has anyone had this problem with Apache and found a solution?

**Comment by JavierMartinz** [comment]
I'm having also this issue and I'm not using a trailing slash in my nginx reverse proxy config

```
  location /_next {
    proxy_pass              http://localhost:3000/_next;
    proxy_http_version      1.1;
    proxy_set_header        Upgrade $http_upgrade;
    proxy_set_header        Connection 'upgrade';
    proxy_set_header        Host $host;
    proxy_cache_bypass      $http_upgrade;
  }

  location /_next/static {
    proxy_pass              http://localhost:3000/_next/static;
    proxy_http_version      1.1;
    proxy_set_header        Upgrade $http_upgrade;
    proxy_set_header        Connection 'upgrade';
    proxy_set_header        Host $host;
    proxy_cache_bypass      $http_upgrade;
  }
```

**Comment by poorscousertommy8** [comment]
I just tested the last version v13.5.3-canary.4, unfortunately the same bug still exists.
@balazsorban44 I'm assuming the label linear:next means you guys are looking at this, am I correct? Can we provide you with further information to support you in some way?

**Comment by jlalmes** [comment]
Been tearing my hair out over this bug today 😭. Was introduced by [`v13.4.13-canary.0`](https://github.com/vercel/next.js/releases/tag/v13.4.13-canary.0) by @ijjk in this PR https://github.com/vercel/next.js/pull/53029.

The problem is that when servers such as NGINX & Apache reverse proxy a request (using `proxy_pass`) they decode the URL before forwarding the request ([StackOverflow](https://serverfault.com/a/906479)). NextJS has always handled these this, but it has been broken since `v13.4.13-canary.0`.

FWIW, this one-liner patch resolves the issue:

<img width="1128" alt="Screenshot 2023-09-25 at 23 09 29" src="https://github.com/vercel/next.js/assets/69924001/b00464fd-d36b-4436-ab21-d4eab8d9cc87">

**Comment by janpaepke** [comment]
@jlalmes thanks for your insights. Would you care adding a pr for this?

**Comment by balazsorban44** [comment]
Hi everyone, we are looking into this!

The best way to help is to provide a **minimal reproduction** instead of "same issue", etc. comments. It should be clear from your reproduction if you are using nginx, standalone (`output: "standalone"`), Docker, or a combination of these. You don't need to comment "it's still happening on x" as this information in itself is not helpful if we cannot see your code/config, instead provide a URL to a public reproduction. :pray:


_Note: Please refrain from tagging maintainers for visibility, the issue is already tracked, so we are aware._

**Comment by omarmciver** [comment]
@balazsorban44 

Minified version of the problem here: [https://github.com/omarmciver/ui-example-web](https://github.com/omarmciver/ui-example-web)

Spin it up with VSCode devcontainer extension. Include simple nginx config.

The README.md give instructions on how to recreate, and fix by not using client side rendering.

Using 13.5.3

**Comment by omarmciver** [comment]
If anyone has a primer or cheat sheet for debugging nextjs src, I'd be happy to get my feet wet. :/ Lazy post... there's a README right?

**Comment by omarmciver** [comment]
Updated the README:
![image](https://github.com/vercel/next.js/assets/56705059/45970c8d-fda7-4ae9-bcc3-a62cc6230c1b)

**Comment by omarmciver** [comment]
## TL;DR :rocket:

If you're using a configuration like this:

```nginx
location /next {
    proxy_pass http://ui-example-web:3000/next;
}
```

Change it to:

```nginx
location /next {
    proxy_pass http://ui-example-web:3000$request_uri;
}
```

:warning: You may also need to add a DNS resolver directive if you don't already have one:

```nginx
resolver 127.0.0.11 valid=30s;
```

## If this helped you, consider [buying me a coffee](https://www.buymeacoffee.com/omarmciver).
---

## Full Explanation :book:

### The Issue :bug:

When NGINX is given a `proxy_pass` value that includes a URI (anything after the protocol, hostname and port), it **always** decodes the URI before forwarding the request.

This leads Next.js to receive a decoded request URI for a static chunk, like so:

```
http://localhost:3000/next/_next/static/chunks/app/things/[thingId]/page-a4112d9d01403386.js
```

It (or webpack?) expects to receive it like so:
```
http://localhost:3000/next/_next/static/chunks/app/things/%5BthingId%5D/page-a4112d9d01403386.js
```

### The Consequence :x:

Next.js doesn't handle that decoded URI well and returns a 404 error. This issue is specific to `next start` (doesn't happen with `next dev`) and occurs only if your dynamic route has a `use client;` directive.

### The Solution :bulb:

The workaround is to use NGINX variables. The `$request_uri` variable contains everything after the original host and port from the incoming request. By appending `$request_uri` at the end of `proxy_pass`, NGINX won't decode the URI.

### Final Thoughts :thought_balloon:

Understanding both NGINX and Next.js behaviors is crucial. I use NGINX in development and NGINX ingress controller in a Kubernetes production environment. The same config change applies to both.

> **Note**: When I switched to this new `proxy_pass` configuration, I had to add a DNS resolver directive to my NGINX config.

---

Feel free to share your thoughts and experiences. If this helped you, I'm still looking for help [buying coffee](https://www.buymeacoffee.com/omarmciver).

**Comment by janpaepke** [comment]
thanks @omarmciver for your detailed analysis.

I would like to point out something from it in particular, because I suspect this to be the case for everyone else having this problem as well.

Initially I thought "Well this again doesn't concern me, because I'm not using nginx, but the nextjs standalone server."
But that quickly changed, when I read this in your text:
> I use NGINX in development and NGINX ingress controller in a Kubernetes production environment.

I don't use nginx in development, but it turns out we DO use an nginx ingress controller in the kubernetes production environment.
I hadn't checked the nginx config on it, because I didn't configure it - I actually think it might have been auto-configured.

**The bad news is**: I have checked the config now and there is no special configuration for `/next` and no `proxy_pass` anywhere, so there's no easy workaround for me. It's also worth noting, that our current configuration worked up until 13.4.12

**The good news is**: We may have found the origin. Next needs to handle the encoding/decoding of the chunk URIs properly. It also means my initial gut feel pointing towards our ingress controller, which I [mentioned in my initial error](https://github.com/vercel/next.js/issues/54008#issuecomment-1707770059) report, was probably correct.

**Comment by poorscousertommy8** [comment]
Sorry Balázs for the tagging.
I would love to provide a minimal reproduction, but I'm having a hard time with it because, similar to my colleagues who use nginx, the problem is probably between Next.js and my Windows Server (as already mentioned, I use iisnode) .

What I do is explained relatively simply: I run the setup according to the Next.js website (npx create-next-app@latest - incl. App Router) and create a folder in the app directory with a dynamic route ([...pagename ]). After deployment, loading the chunks fails.

**Comment by omarmciver** [comment]
@janpaepke I can share with you the Kubernetes ingress solution a little later today. I don't know what, if any, blocker there is for nextjs handling a decoded uri. I might look at that previous commit that was referenced as introducing the issue.

**Comment by omarmciver** [comment]
@janpaepke  and others.
I didn't spend time on my AKS ingress config. Instead, I managed to get up and running with debugging nextjs and I've got a fix in a draft PR.

PR #56187 

It will remain in draft until I get time to write a test to support this.

**Comment by alexrabin** [comment]
@omarmciver 
<img src="https://media1.giphy.com/media/okLCopqw6ElCDnIhuS/giphy.gif?cid=ecf05e47nxfrijthsfz69lmg3brj68en75hijmvm7zr0bk01&ep=v1_gifs_search&rid=giphy.gif&ct=g" />

**Comment by jviall** [comment]
> @janpaepke and others. I didn't spend time on my AKS ingress config. Instead, I managed to get up and running with debugging nextjs and I've got a fix in a draft PR.
> 
> PR #56187
> 
> It will remain in draft until I get time to write a test to support this.

Tysm for your efforts and contributions!!

**Comment by omarmciver** [comment]
Still working on getting a test in place, then hopefully working my way to the top of the maintainers review list. Man, they have a lot of work!

**Comment by poorscousertommy8** [comment]
For all users who use windows and iisnode, there is a solution https://github.com/vercel/next.js/issues/54325#issuecomment-1742153571.

**Comment by janpaepke** [comment]
For all followers of this: There's movement in @omarmciver's [PR](https://github.com/vercel/next.js/pull/56187). Looks like this might be solved soon. 🥳

**Comment by huozhi** [comment]
This is fixed in latest patch release 13.5.4, please upgrade to the new version, thanks y'all for digging deep on this issue 🙏

**Comment by mick-feller** [comment]
i can confirm that this patch worked, i upgraded to 13.5.5 and all good again!

**Comment by sakib412** [comment]
> This is fixed in latest patch release 13.5.4, please upgrade to the new version, thanks y'all for digging deep on this issue 🙏

Thanks, fixed in 13.5.4 I was in 13.5.3

**Comment by rihards-simanovics** [comment]
people please stop spamming, this is not a discussion but an issue, and I only want to get updates when there is progress, if it works for you just 👍the first message that says it started working or the issue got fixed.
